### ISSUE TITLE
Call to undefined function Symfony\Component\DependencyInjection\Loader\Configurator\service()

### ISSUE THREAD

**Issue by tomasfejfar** [rectorphp/rector#5083]
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | rector/rector:latest             |
| Installed as   | dockerimage                          |

```
λ docker run -it --rm --entrypoint bash -v %CD%:/project rector/rector:latest
root@440c45a4a3fb:/project# cd legacy-app
root@440c45a4a3fb:/project/legacy-app# which rector
/rector/bin/rector
root@440c45a4a3fb:/project/legacy-app# rector process . --dry-run -vvvv


 [ERROR] Call to undefined function Symfony\Component\DependencyInjection\Loader\Configurator\service()


root@440c45a4a3fb:/project/legacy-app# rector --version


 [ERROR] Call to undefined function Symfony\Component\DependencyInjection\Loader\Configurator\service()

```

```
<?php

declare(strict_types=1);

use Rector\Core\Configuration\Option;
use Rector\Php74\Rector\Property\TypedPropertyRector;
use Rector\Set\ValueObject\SetList;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    // get parameters
    $parameters = $containerConfigurator->parameters();

    // Define what rule sets will be applied
    $parameters->set(Option::SETS, [
        'php72',
    ]);

    // get services (needed for register a single rule)
    // $services = $containerConfigurator->services();

    // register a single rule
    // $services->set(TypedPropertyRector::class);
};

```

## Minimal PHP Code Causing Issue

No idea where to start the debug. It seems that rector does not output stacktrace for its exceptions even with `-vvv`

## Expected Behaviour

This does not happen :)

**Comment by TomasVotruba** [comment]
Thanks for reporting. Try `--debug` option.
That looks like Docker autoloading issue.

If you include the function manually in composer.json files, does it help?

It's here:
https://github.com/symfony/dependency-injection/blob/5.x/Loader/Configurator/ContainerConfigurator.php#L109

Docker might be crappy here. How does [prefixed version do](https://github.com/rectorphp/rector-prefixed)? If PHP 7.3 is not a blocker, give it a try. Might be easier to use.

**Comment by tomasfejfar** [comment]
```
root@cfd7f4f3afdb:/project/legacy-app# rector --debug


 [ERROR] Call to undefined function Symfony\Component\DependencyInjection\Loader\Configurator\service()

```

**Comment by tomasfejfar** [comment]
```
  Package rector/rector-prefixed at version  has a PHP requirement incompatible with your PHP version (7.1.30)
```

Intention was to use it to migrate from 7.1.30 to something of late :) It's why I went with docker in the first place.

**Comment by TomasVotruba** [comment]
What about this one?

> If you include the function manually in composer.json files, does it help?

> It's here:
https://github.com/symfony/dependency-injection/blob/5.x/Loader/Configurator/ContainerConfigurator.php#L109

**Comment by tomasfejfar** [comment]
How and where would I do that? Should I add it in my composer file in my project? Or in rector's composer inside the docker image?

**Comment by tomasfejfar** [comment]
(JFYI the latest pushed tag 0.8.8 docker image works, but is missing for example the init command, etc. )

**Comment by TomasVotruba** [comment]
> How and where would I do that?

[Like this](https://getcomposer.org/doc/04-schema.md#files)

> Should I add it in my composer file in my project?

Yes.

> (JFYI the latest pushed tag 0.8.8 docker image works, but is missing for example the init command, etc. )

The latest GitHub tag is: https://github.com/rectorphp/rector/releases/tag/0.9.4
I'll push a new tag, maybe it needs a new tag ping :)

**Comment by TomasVotruba** [comment]
Btw, there might be one more easier way. This it not official yet (so don't tell anybody please :))

We have a WIP PHP 7.1 version Rector in here: https://github.com/rectorphp/rector-php71
@leoloso is behind it and I have information it should be working. You can give it a try and reach out

**Comment by tomasfejfar** [comment]
Both version (0.9.5 and latest) docker images work now, thanks!

**Comment by tomasfejfar** [comment]
Yea, so docker won't work either as long as it is prefixed docker :) 

```
Fatal error: Declaration of Service_AuditLogEventDispatcher::dispatch($event, ?string $eventName = NULL) must be compatible with Symfony\Component\EventDispatcher\EventDispatcher::dispatch(object $event, ?string $eventName = NULL): object
```

I can't match that because the app is 7.1. It can't have object typehint. So I'll give the 7.1 version a try.

**Comment by TomasVotruba** [comment]
Yea, these kinds of problems are the reason we go for scoped & downgraded package version. Also it's much easier to add it and debug it.

**Comment by tomasfejfar** [comment]
Tried it, I guess the 7.1 version would need to be scoped as well to work :'( 

Although for reasons yet unknown it breaks on symfony iself:
```
PHP Fatal error:  Declaration of Symfony\Component\Cache\CacheItem::tag($tags): Symfony\Component\Cache\CacheItem must be compatible with Symfony\Contracts\Cache\ItemInterface::tag($tags): Symfony\Contracts\Cache\ItemInterface in /var/www/html/legacy-app/vendor/rector/rector-php71/vendor/symfony/cache/CacheItem.php on line 22
```

IMHO scoping + docker is a way to go. There would be no conflicts whatsoever.

**Comment by TomasVotruba** [comment]
> Tried it, I guess the 7.1 version would need to be scoped as well to work :'(

Oh, I forgot about that :/ @leoloso Could you incorporate that into the build process? It's not helpful without that.

> IMHO scoping + docker is a way to go. There would be no conflicts whatsoever.

Still much slower than without Docker and harder to debug. The Docker path paradigm add huge cognitive overload.

**Comment by zajca** [comment]
Hi,
I made some digging into this (I'm working with Tomáš on same code). 
We manage to use 0.8.8 docker image to update codebase to PHP 7.3 :tada: 

But now I'm kind of stuck.
- On 0.8.8, I can't use config file, when I do I always get:
`[ERROR] Undefined class constant 'RESERVED_CHARACTERS'`
RESERVED_CHARACTERS constant is part of `Symfony\Contracts\Cache` but there no such thing in our codebase and it's nowhere used so this is probably from docker image it self (I don't expect this to be fixed, it's old version)

- On 0.9.5 I get same error as Tomáš, but we are using symfony DI ^4.4 and there is not such thing as `Symfony\Component\DependencyInjection\Loader\Configurator\service()` so we can't autoload it.

I guess that rector is somehow mixing own dependencies with the ones in project dir probably when `autoload-file` is used from project.

I would use prefixed version but that conflicts with `phpstan-shim` and upgrade is bit more difficult.

**Comment by TomasVotruba** [comment]
Hi,

the `phpstan-shim` is not needed anymore, as phpstan is prefixed out of the box.
Try the prefixed Rector with `phpstan/phpstan`

**Comment by leoloso** [comment]
This error:

```
PHP Fatal error:  Declaration of Symfony\Component\Cache\CacheItem::tag($tags): Symfony\Component\Cache\CacheItem must be compatible with Symfony\Contracts\Cache\ItemInterface::tag($tags): Symfony\Contracts\Cache\ItemInterface in /var/www/html/legacy-app/vendor/rector/rector-php71/vendor/symfony/cache/CacheItem.php on line 22
```

was happening because returning `self` in PHP 7.1 doesn't work (check #5145), and those functions [return `self`](https://github.com/symfony/symfony/blob/ff22d30cb02fad4a3e97245d46411cea05fcb36b/src/Symfony/Contracts/Cache/ItemInterface.php#L57).

I just submitted a PR that should fix that issue: #5146.

---

Should Rector in PHP 7.1 be scoped? 

- If `rector/rector-php71` is executed as a standalone, then there's no need since it's the only one loading classes
- If `rector/rector-php71` is added to a project's `composer.json`, then yes, it's needed

So I guess we must implement scoping to be on the safe side. But we could also add instructions that `rector/rector-php71` is to be used as a standalone, otherwise `rector/rector` must be used. Both options should be acceptable.

**Comment by TomasVotruba** [comment]
Closing as original issue is resolved
